src/model/differ.js

```diff
@@ -0,0 +1,774 @@
+
+						for ( const oldChange of changes ) {
+							this._handleChange( changeItem, oldChange, changes );
+						}
+
+						changes.push( changeItem );
+					}
+
+					inc.howMany = old.offset - inc.offset;
+				} else if ( inc.offset >= old.offset && inc.offset < oldEnd ) {
+					if ( incEnd > oldEnd ) {
+						inc.howMany = incEnd - oldEnd;
+						inc.offset = oldEnd;
+					} else {
+						inc.howMany = 0;
+					}
+				}
+			}
+
+			if ( old.type == 'attribute' ) {
+				if ( inc.offset >= old.offset && incEnd <= oldEnd ) {
+					inc.howMany = 0;
+				}
+			}
+		}
+	}
+
+	/**
+	 * Returns an object with a single insert change description.
+	 *
+	 * @private
+	 * @param {module:engine/model/element~Element} parent Element in which change happened.
+	 * @param {Number} offset Offset at which change happened.
+	 * @param {String} name Removed element name or `'$text'` for character.
+	 * @returns {Object} Diff item.
+	 */
+	_getInsertDiff( parent, offset, name ) {
+		return {
+			type: 'insert',
+			position: Position.createFromParentAndOffset( parent, offset ),
+			name,
+			length: 1,
+			changeCount: this._changeCount++
+		};
+	}
+
+	/**
+	 * Returns an object with a single remove change description.
+	 *
+	 * @private
+	 * @param {module:engine/model/element~Element} parent Element in which change happened.
+	 * @param {Number} offset Offset at which change happened.
+	 * @param {String} name Removed element name or `'$text'` for character.
+	 * @returns {Object} Diff item.
+	 */
+	_getRemoveDiff( parent, offset, name ) {
+		return {
+			type: 'remove',
+			position: Position.createFromParentAndOffset( parent, offset ),
+			name,
+			length: 1,
+			changeCount: this._changeCount++
+		};
+	}
+
+	/**
+	 * Returns an array of objects that each is a single attribute change description.
+	 *
+	 * @private
+	 * @param {module:engine/model/range~Range} range Range on which change happened.
+	 * @param {Map} oldAttributes Map, map iterator or compatible object that contains attributes before change.
+	 * @param {Map} newAttributes Map, map iterator or compatible object that contains attributes after change.
+	 * @returns {Array.<Object>} Array containing one or more diff items.
+	 */
+	_getAttributesDiff( range, oldAttributes, newAttributes ) {
+		// Results holder.
+		const diffs = [];
+
+		// Clone new attributes as we will be performing changes on this object.
+		newAttributes = new Map( newAttributes );
+
+		// Look through old attributes.
+		for ( const [ key, oldValue ] of oldAttributes ) {
+			// Check what is the new value of the attribute (or if it was removed).
+			const newValue = newAttributes.has( key ) ? newAttributes.get( key ) : null;
+
+			// If values are different (or attribute was removed)...
+			if ( newValue !== oldValue ) {
+				// Add diff item.
+				diffs.push( {
+					type: 'attribute',
+					position: range.start,
```

Comment:
It's strange that I get both position and range.

Reply:
`position` is useful for further sorting (as other changes has `position` too), but I will remove it from the final results. BTW. attribute change item also has `length` which is not used at all - I will remove it.

Reply:
It appears that I was wrong, both `length` and `position` are used in further processing but also both are not really needed later so they can be removed before changes are returned by `getChanges()`